### ISSUE TITLE
Rule proposal SX1027: IndentUsingTabs

### ISSUE THREAD

**Issue by AArnott** [DotNetAnalyzers/StyleCopAnalyzers#730]
As found [here](https://github.com/AArnott/nerdbank.stylecop.rules/blob/master/src/NerdBank.StyleCop.Rules/NerdBankRules.cs) and described [here](http://blog.nerdbank.net/2008/09/notrailingwhitespace-stylecop-rule-and.html), the IndentUsingTabs rule is the inverse of SA1027 and serves those folks who prefer tabs over spaces.

This issue tracks porting the rule from the legacy StyleCop implementation. You may assign the issue to me.

As part of this proposal, we should consider merging my OneTabIndent rule into SX1027:

OneTabIndent sub-rule

Consider the following code, and imagine tab width + indent are set to 5 characters for sake of conveniently making these characters "visible" in the below snippet:

```csharp
<TAB>var foobar = select a
<TAB><TAB><TAB>   from source;
```

This is roughly what VS would do by default when using tabs. It uses as many tabs as possible to intend the second line then falls back to spaces for the remainder of the indent in order for the LINQ query to line up with the previous line. While this looks good with tabs set to 5 characters wide it will *not* look good when the IDE on another user's machine is configured for 3 characters:

```csharp
TABvar foobar = select a
TABTABTAB   from source;
```

This of course is one of the chief arguments against using tabs. But this is a fault of the language service's use of mixed tabs and spaces rather than a fundamental flaw with tabs. If instead the original characters were:

```csharp
<TAB>var foobar = select a
<TAB>             from source;
```

Then again when 5 column tabs are changed to 3 column, the appearance is still good:

```csharp
TABvar foobar = select a
TAB             from source;
```

This proposed rule is to enforce and preserve the optimal combination of tabs and spaces so that it looks good at any tab width.

I actually had convinced the Roslyn team to fix their default behavior for indentation to produce the correct combination of tabs and spaces since they were writing it from scratch, although I haven't confirmed whether they do it or not this close to RTM.

**Comment by bmenees** [comment]
I also prefer tabs for indentation, and I missed the StyleCop+ SP2001: CheckAllowedIndentationCharacters rule that I previously used to require tabs.  So I wrote my own [MEN001: TabsShouldBeUsed] (http://www.menees.com/Analyzers.htm) rule in my new Menees.Analyzers assembly.  Perhaps it will help you (and others) out.

**Comment by sharwell** [comment]
Superseded by configuration-based proposal in #2035, #2037.

@AArnott I haven't proposed the OneTabIndent rule directly because it's non-trivial to describe. For example, I consider each of the following to be the desired indentation:

```csharp
<TAB>var foobar = select a
<TAB>             from source;

<TAB>var foobar =
<TAB><TAB>select a
<TAB><TAB>from source;

<TAB>int[] foobar =
<TAB><TAB>new[]
<TAB><TAB>{
<TAB><TAB><TAB>3
<TAB><TAB>};
```

I believe it's fairly straightforward to describe the rules at a high level:

> Suppose you have lines *x* and *y*, where *x* > *y* and the indentation of *x* is relative to position *p* within line *y*.
> * The number of tabs used in *x* to reach position *p* should be equal to the number of leading tab characters on line *y* preceding position *p*. The remaining characters used in *x* to reach position *p* should be space characters.
> * If *x* needs to be indented one tab stop further than *p* due to indentation rules, the additional indentation should be a single tab character if the indentation to reach *p* was entirely tab characters. Otherwise, the additional indentation should be `indentationSize` number of space characters.

However, the implementation may prove more challenging than this so I'm not sure I'm ready to make a formal proposal. If you want to take a crack at it as a new issue, please feel free to do so.
